### PR TITLE
Restructuring make script

### PR DESCRIPTION
**Pull request by Jiyuu**
In addiction to restructuring I have also made the "all" command fetch dependencies.

I believe this structure is more readable and obvious, but if you think this structure is not proper for powershell LMK and I'll make a separate PR just for adding the dependencies to the "all" command.

### CODE REVIEW

**Comment by pchote** [comment]
Can some of our windows-based people please review this?

File: make.ps1 [hunk 1]
```diff
@@ -1,27 +1,5 @@
...
-		}
-	}
-	return $null
-}
-
-function UtilityNotFound
-{
-	echo "OpenRA.Utility.exe could not be found. Build the project first using the `"all`" command."
-}
+$functions = "$($MyInvocation.MyCommand.path | split-path)\make-functions.ps1"
```
Comment (pchote) [inline make.ps1]: I'm not convinced that the (minor, IMO?) readability improvement for shipping these into their own file outweighs the (minor) complexity of having a separate function file.  Wouldn't we get the same benefits by having all the functions at the top of the file?

Comment (Jiyuu) [inline make.ps1]: You will get most of the benefit, but it would take a few more seconds figuring out that the heart of the code is at the bottom (unless this is some kind of an agreed practice in powershell i am simply unaware of)

Basically I tried making the file as accessible as possible for someone opening it for the first time.


Though if there's strong agreement that its better not to have 2 files I don't mind combining them.

Comment (abcdefg30) [inline make.ps1]: I'd like to have them stay in one file.

**Comment by Jiyuu** [comment]
I've combined the functions back into the original file and added some comments in the hope this will be clear enough.

Any comments?

File: make.ps1 [hunk 2]
```diff
@@ -225,20 +186,68 @@ elseif ($command -eq "docs")
...
+		{
+			return $path
 		}
-		Start-Sleep -Milliseconds 50
+	}
+	return $null
+}
+
+
+
```
Comment (abcdefg30) [inline make.ps1]: I think we can reduce this three new lines to one, and also add one new line below the `UtilityNotFound` function.

File: make.ps1 [hunk 3]
```diff
@@ -186,13 +143,16 @@ elseif ($command -eq "check")
 		./OpenRA.Utility.exe cnc --check-code-style OpenRA.Utility
 		echo "Checking for code style violations in OpenRA.Test..."
 		./OpenRA.Utility.exe cnc --check-code-style OpenRA.Test
+		echo "Checking for explicit interface violations..."
+		./OpenRA.Utility.exe all --check-explicit-interfaces
```
Comment (abcdefg30) [inline make.ps1]: Why is this moved around?

File: make.ps1 [hunk 4]
```diff
@@ -225,20 +186,68 @@ elseif ($command -eq "docs")
...
+	echo "  docs            Generates the trait and Lua API documentation."
+	echo ""
+	$command = (Read-Host "Enter command").Split(' ', 2)
+}
+else
+{
+	$command = $args
+}
+
+switch ($command) {
```
Comment (abcdefg30) [inline make.ps1]: Put the brace on a new line, please.

File: make.ps1 [hunk 5]
```diff
@@ -225,20 +186,68 @@ elseif ($command -eq "docs")
...
+	"all" { All-Command }
+	"dependencies" { Dependencies-Command }
+	"version" { Version-Command }
+	"clean" { Clean-Command }
+	"test" { Test-Command }
+	"check" { Check-Command }
+	"check-scripts" { Check-Scripts-Command }
+	"docs" { Docs-Command }
+	Default {
```
Comment (abcdefg30) [inline make.ps1]: Same here (or put everything on one line: `Default { echo (...) }`).

File: make.ps1 [hunk 6]
```diff
@@ -1,52 +1,10 @@
...
-	echo "  check-scripts   Checks .lua files for syntax errors."
-	echo "  docs            Generates the trait and Lua API documentation."
-	echo ""
-	$command = (Read-Host "Enter command").Split(' ', 2)
-}
-else
-{
-	$command = $args
-}
+####### The starting point for the script is the bottom #######
```
Comment (abcdefg30) [inline make.ps1]: Why actually? Couldn't we just have the main part at the top?

File: make.ps1 [hunk 7]
```diff
@@ -1,52 +1,10 @@
...
 
-if ($command -eq "all")
-{
+###############################################################
+########################## FUNCTIONS ##########################
+###############################################################
```
Comment (abcdefg30) [inline make.ps1]: This looks like a bit of overkill to me, but I don't really mind that much either. Maybe others have stronger opinions.

File: make.ps1 (hunk 6 above)
Comment (Jiyuu) [inline make.ps1]: Nope
Apparently Powershell doesn't support forward declaration, this is the reason I went with the separate file in the first place.

Comment (abcdefg30) [inline make.ps1]: Ok. I already feared that. This is fine then.

File: make.ps1 (hunk 3 above)
Comment (Jiyuu) [inline make.ps1]: not sure, i see that this is the exact opposite of the last commit before me, so either i used an old version by mistake, or alternatively i remember seeing some errors when i was testing all the commands so i may have fixed something and forgot about it. 

will return back for now

File: make.ps1 [hunk 8]
```diff
@@ -1,52 +1,10 @@
...
-{
-	$command = $args
-}
-
-if ($command -eq "all")
-{
+###############################################################
+########################## FUNCTIONS ##########################
+###############################################################
+function All-Command {
```
Comment (abcdefg30) [inline make.ps1]: The brace here also wants to go on a new line (likewise for `Clean-Command`, `Check-Command` and `Version-Command` below).

**Comment by abcdefg30** [comment]
Looks good to me now (I'm even ok with adding the `dependencies` step to `all`). Just two things:
- The window now immediately closes instead of waiting for key input like it did before
- Please squash the commits

**Comment by Jiyuu** [comment]
I restored the piece for keeping the window open when the script is called without any parameters

**Comment by abcdefg30** [comment]
Thanks. 👍 / ✅

**Comment by reaperrr** [comment]
> Looks good to me now (I'm even ok with adding the dependencies step to all)

Same here 👍
